### PR TITLE
Llvm15

### PR DESCRIPTION
**Pull request by thewilsonator**

### CODE REVIEW

File: gen/asmstmt.cpp [hunk 1]
```diff
@@ -694,6 +694,9 @@ void CompoundAsmStatement_toIR(CompoundAsmStatement *stmt, IRState *p) {
   std::vector<LLType *> types;
   types.insert(types.end(), outtypes.begin(), outtypes.end());
   types.insert(types.end(), intypes.begin(), intypes.end());
+  //FIXME: LLVM 15 requires `elementtype` attribute for indirect constraints (i.e. "=*m)
+  // e.g.  Operand for indirect constraint must have elementtype attribute
+  // call void asm sideeffect "stp x19, x20, $0", "=*m,=*m"(i64* %6, i64* %7),
```
Comment (thewilsonator) [inline gen/asmstmt.cpp:699]: @kinke I'm currently stuck on this. This should be 
```
call void asm sideeffect "stp x19, x20, $0", "=*m,=*m"(i64* elementtype(i64) %6, i64* elementtype(i64) %7)
```

How are parameter attributes added?

Comment (thewilsonator) [reply gen/asmstmt.cpp]: Fixed this

File: gen/llvmhelpers.cpp [hunk 2]
```diff
@@ -1783,6 +1783,8 @@ llvm::GlobalVariable *declareGlobal(const Loc &loc, llvm::Module &module,
   llvm::GlobalVariable *existing =
       module.getGlobalVariable(mangledName, /*AllowInternal=*/true);
   if (existing) {
+#if LDC_LLVM_VER < 1500
+    //FIXME: LLVM 15 removed getElementType()
     const auto existingType = existing->getType()->getElementType();
```
Comment (kinke) [inline gen/llvmhelpers.cpp:1788]: According to https://llvm.org/docs/OpaquePointers.html#migration-instructions, there should be some `getValueType()` for globals.

File: gen/asmstmt.cpp [hunk 3]
```diff
@@ -721,6 +721,18 @@ void CompoundAsmStatement_toIR(CompoundAsmStatement *stmt, IRState *p) {
...
       ia, args, retty == LLType::getVoidTy(gIR->context()) ? "" : "asm");
+#if LDC_LLVM_VER >= 1500
+  {
+    int i = 0;
+    for (llvm::InlineAsm::ConstraintInfo& CI : ia->ParseConstraints()) {
+      if (CI.isIndirect)
+        call->addParamAttr(i, llvm::Attribute::get(getGlobalContext(),
+                                                   llvm::Attribute::ElementType,
+                                                   types[i]));
+        i++;
```
Comment (kinke) [inline gen/asmstmt.cpp:732]: Wrong indentation.

File: gen/asmstmt.cpp [hunk 4]
```diff
@@ -721,6 +721,18 @@ void CompoundAsmStatement_toIR(CompoundAsmStatement *stmt, IRState *p) {
...
   llvm::CallInst *call = p->ir->CreateCall(
       ia, args, retty == LLType::getVoidTy(gIR->context()) ? "" : "asm");
+#if LDC_LLVM_VER >= 1500
+  {
+    int i = 0;
+    for (llvm::InlineAsm::ConstraintInfo& CI : ia->ParseConstraints()) {
+      if (CI.isIndirect)
+        call->addParamAttr(i, llvm::Attribute::get(getGlobalContext(),
+                                                   llvm::Attribute::ElementType,
+                                                   types[i]));
```
Comment (kinke) [inline gen/asmstmt.cpp:731]: Is `types[i]` really the *pointee* type here?

File: gen/passes/DLLImportRelocation.cpp [hunk 5]
```diff
@@ -176,6 +176,18 @@ struct Impl {
...
+               "Replacing operand with value of different type!");
+        if (ce->getOperand(OpNo) == Op)
+          return const_cast<ConstantExpr*>(ce);
+
+        SmallVector<Constant*, 8> NewOps;
+        for (unsigned i = 0, e = ce->getNumOperands(); i != e; ++i)
+          NewOps.push_back(i == OpNo ? Op : ce->getOperand(i));
+
+        return ce->getWithOperands(NewOps);
+  }
```
Comment (kinke) [inline gen/passes/DLLImportRelocation.cpp:190]: The assert+identity checks are superfluous for our single use below; should probably be doable in 3 lines - copy operands, replace first entry, call `getWithOperands()`.

File: gen/passes/GarbageCollect2Stack.cpp [hunk 6]
```diff
@@ -199,8 +199,10 @@ class ArrayFI : public TypeInfoFI {
     assert(ArrTy && "Dynamic array type not a struct?");
     assert(isa<IntegerType>(ArrTy->getElementType(0)));
     const PointerType *PtrTy = cast<PointerType>(ArrTy->getElementType(1));
+#if LDC_LLVM_VER < 1500
+    //FIXME for LLVM 15
     Ty = PtrTy->getElementType();
```
Comment (kinke) [inline gen/passes/GarbageCollect2Stack.cpp:204]: Should still be hackable via `PtrTy->getPointerElementType()`, that's what `getPointeeType()` does anyway.

File: gen/tocall.cpp [hunk 7]
```diff
@@ -93,12 +93,12 @@ LLValue *DtoCallableValue(DValue *fn) {
...
 ////////////////////////////////////////////////////////////////////////////////
 
-LLFunctionType *DtoExtractFunctionType(LLType *type) {
+LLFunctionType *DtoExtractFunctionType(LLType *type, LLValue *val) {
   if (LLFunctionType *fty = isaFunction(type)) {
     return fty;
   }
   if (LLPointerType *pty = isaPointer(type)) {
-    if (LLFunctionType *fty = isaFunction(pty->getElementType())) {
+    if (LLFunctionType *fty = isaFunction(getPointeeType(val))) {
```
Comment (kinke) [inline gen/tocall.cpp:101]: Ditto here, then you don't need a signature change.

File: ir/irclass.cpp [hunk 8]
```diff
@@ -109,6 +109,8 @@ LLGlobalVariable *IrClass::getClassInfoSymbol(bool define) {
       // regular TypeInfo metadata
       emitTypeInfoMetadata(typeInfo, aggrdecl->type);
 
+//FIXME: LLVM15 this probably needs a refactor of DtoType for classes because getElementType() no longer exists
+#if LDC_LLVM_VER < 1500
       // Gather information
       LLType *type = DtoType(aggrdecl->type);
       LLType *bodyType = llvm::cast<LLPointerType>(type)->getElementType();
```
Comment (kinke) [inline ir/irclass.cpp:116]: And here - `type->getPointerElementType()`.

File: gen/passes/GarbageCollect2Stack.cpp (hunk 6 above)
Comment (thewilsonator) [reply gen/passes/GarbageCollect2Stack.cpp:204]: The Migration instructions say:
```
In order to support opaque pointers, two types of changes tend to be necessary. The first is the removal of all calls to PointerType::getElementType() and Type::getPointerElementType().
```

So it seems we should not be using those either.

File: gen/tocall.cpp (hunk 7 above)
Comment (thewilsonator) [reply gen/tocall.cpp:101]: Ditto

File: ir/irclass.cpp (hunk 8 above)
Comment (thewilsonator) [reply ir/irclass.cpp:116]: ditto

File: gen/passes/GarbageCollect2Stack.cpp (hunk 6 above)
Comment (kinke) [reply gen/passes/GarbageCollect2Stack.cpp:204]: We are *all over the place* via `getPointeeType()` anyway. Properly migrating to opaque pointers will take some work.

**Pull request closed without merging**